csharp/source/autowrap/csharp/csharp.d

```diff
@@ -0,0 +1,1061 @@
+module autowrap.csharp.csharp;
+
+import autowrap.csharp.boilerplate;
+import autowrap.csharp.common;
+import autowrap.reflection;
+
+import std.ascii;
+import std.datetime;
+import std.traits;
+import std.string;
+import std.uuid;
+
+import std.stdio;
+
+private __gshared string[string] returnTypes;
+private __gshared csharpNamespace[string] namespaces;
+private __gshared csharpRange rangeDef;
+
+enum string voidTypeString = "void";
+enum string stringTypeString = "string";
+enum string wstringTypeString = "wstring";
+enum string dstringTypeString = "dstring";
+enum string boolTypeString = "bool";
+enum string dateTimeTypeString = "DateTime";
+enum string sysTimeTypeString = "SysTime";
+enum string uuidTypeString = "UUID";
+enum string charTypeString = "char";
+enum string wcharTypeString = "wchar";
+enum string dcharTypeString = "dchar";
+enum string ubyteTypeString = "ubyte";
+enum string byteTypeString = "byte";
+enum string ushortTypeString = "ushort";
+enum string shortTypeString = "short";
+enum string uintTypeString = "uint";
+enum string intTypeString = "int";
+enum string ulongTypeString = "ulong";
+enum string longTypeString = "long";
+enum string floatTypeString = "float";
+enum string doubleTypeString = "double";
+enum string sliceTypeString = "slice";
+
+private class csharpNamespace {
```

Comment:
camelCase for class names?

Why `class`? I see no dynamic polymorphism.

Reply:
Because .require doesn't work with structs. See: getNamespace and getAggregate

Reply:
The only thing that only works with classes are dynamic polymorphism and subtyping:

```d
struct Foo { int i; }

Foo[string] stringToFoo;

void main() {
    stringToFoo.require("foo", Foo(42));
    assert(stringToFoo["foo"] == Foo(42));
}
```

Reply:
That does not work in practice because structs are copied so what in effect happens is that all the strings that I put in there are lost and string.empty gets emitted because the copy of the struct in the AA is never updated. I've tried it twice now.

Reply:
The value type of the AA can be changed to Struct* instead of Struct. I still don't see a reason to use classes.

Reply:
...
wat

How is littering the code with raw pointers better than a class? That is going to take a LOT of explaining to convince me.

Reply:
How is a class reference different from a struct pointer in anything but name and syntax?

In C++, `class` can mean anything. In D, `class` means there's polymorphism involved, otherwise it's just confusing. That's the only reason D classes exist.

Reply:
You are 100% technically correct. To the computer it makes no difference. But to the humans maintaining the code, `Struct*` is significantly more error prone. At some point in the future when somebody else is maintaining this code they will run headlong into this and the frustration and lost time it causes will immediately outweigh any benefit. Speaking of which, more specifically, what is the _benefit_ of a struct pointer over a class?

Just because we can do something does not mean that we should do something. Programming is the art of making a computer do things with code; engineering is the art of making the computer do things with code in a way that does not encourage the human maintainers to make mistakes later. (ex: OpenSSL)

Reply:
> Struct* is significantly more error prone.

I don't see how it is. It's as error prone as `Class`, since they're both pointers.

> At some point in the future when somebody else is maintaining this code they will run headlong into this and the frustration and lost time it causes will immediately outweigh any benefit

We obviously want to avoid that, but I don't see how struct pointers are different from class references in that regard. Do you have an example of what you mean?

> Speaking of which, more specifically, what is the benefit of a struct pointer over a class?

Clarity - classes signal the intention to store them as a base interface/abstract class together and call virtual member functions on them.

> Just because we can do something does not mean that we should do something. 

Agreed. 

> Programming is the art of making a computer do things with code; engineering is the art of making the computer do things with code in a way that does not encourage the human maintainers to make mistakes later. (ex: OpenSSL)

Also agreed.

I'm having difficulty understanding what problems you think the codebase would have with struct pointers instead of class references.

Reply:
Your list of what classes provide is correct but incomplete. What _else_ do classes provide out of the box?
Reference Semantics.
You are asking me to (poorly) recreate something that already exists in the language. You are asking me to litter the code with (*Struct).value C-style dereferencing code.
D isn't C++ where classes can be value or reference, in D classes always mean reference semantics. In this case I explicitly want reference semantics, so I use the built-in language tool. I use the built in language tool because it is guaranteed to always work the way I need it and nobody else is going to make frustrating mistakes working with it.

You asking to me to not use the built-in language tool. The tool that everyone, everywhere, understands how works. The tool that is guaranteed by the language to work exactly how I need it to. But most importantly, the tool that is designed and intended to do this. 

This is an extraordinary ask, and therefore demands extraordinary reasoning. Simply saying that there is no polymorphism or sub-typing isn't enough, there needs to be a significant problem with using classes.

There isn't. I remain entirely unconvinced.

Reply:
> You are asking me to litter the code with (*Struct).value C-style dereferencing code.

That's only ever true for operators (which is silly and should be fixed). Even then, given a struct pointer `foo`, `foo.opSlice` works. Other than those, why would one ever write `(*Struct).value` instead of `Struct.value`? D deferences automatically and makes one wonder why `->` even exists in C.

> The tool that everyone, everywhere, understands how works

I've never used classes for reference semantics. It's never even occurred to me. Being used that way would prompt me to add (or ask to be added) a comment why classes were being used with no virtual member function call in sight. And make them `final`.

Reply:
If we can agree that "It didn't occur to me" is not a reasonable critique of the code, then I think we can move forward.

Here is the thing, if I saw raw pointers in D code it would absolutely prompt me to start asking questions about why they are being used. Raw pointers in general have caused more damage to the world than any other single feature of a programming language. With that in mind it is my opinion that raw pointers are a tool of last resort, too be used when no other method of accomplishing the goal exists. Since D has classes with enforced reference semantics I will 100% of the time attempt to use those before I use raw pointers when I need reference semantics on an aggregate.

Making them final is fine. No problem with that.

Reply:
Let's compromise with making them `final` and adding a comment explaining why they're classes.

I still don't understand (because I haven't seen any code) why you think "raw" pointers are in any way, shape, or form more dangerous than class references in D. There's no semantic difference between them that I can see. Am I missing something? What's the difference in language enforcement between:

```d
void myfunc(ClassType obj);
```

and

```d
void myfunc(StructType* obj);
```

?

They both suffer from the same billion-dollar mistake.

Would you feel more comfortable with a [NotNull!(StructType*)](https://github.com/aliak00/optional)?

I just don't see what the deal is. In C and C++ raw pointers are bad because of lifetimes. In D one can just `new` a struct and it's never going to be an issue.

Reply:
Work for me.
My overall point is that I am not a fan of using non-standard techniques to achieve something that is entirely covered by existing language features. I won't argue that it is _technically_ equivalent, just that it's a non-standard when something that is standard is readily available to achieve the goal.

Reply:
I'm not a fan of that either but we seem to disagree on what is standard here. Anyway, final classes with comments works for me.